File: WdtcCore/src/main/java/org/wdt/wdtc/launch/LaunchGame.java
package org.wdt.wdtc.launch;

import lombok.SneakyThrows;
import org.apache.log4j.Logger;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.download.DownloadGameVersion;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.manger.FileManger;
import org.wdt.wdtc.utils.WdtcLogger;

import java.io.File;
import java.io.IOException;

public class LaunchGame {
    private static final File StartBat = FileManger.getStarterBat();
    private static final Logger logmaker = WdtcLogger.getLogger(LaunchGame.class);
    private final Launcher launcher;


    public LaunchGame(Launcher launcher) {
        this.launcher = launcher;
        try {
            launcher.LaunchTask();
            logmaker.info("Start Download");
            DownloadGameVersion gameVersion = new DownloadGameVersion(launcher);
            try {
                gameVersion.DownloadGame();
            } catch (Throwable e) {
                logmaker.error("", e);
            }
            logmaker.info("Downloaded Finish");
            logmaker.info("Write Start Script");
            writeStartScript();
            logmaker.info("Write Start Script Finish");
            logmaker.info("Start Run Start Script,Console:" + launcher.Console());
            logmaker.info(String.format("Launch Version: %s - %s", launcher.getVersionNumber(), launcher.getKind().toString()));
            logmaker.info(launcher.getGameConfig().getConfig());
        } catch (Exception e) {
            logmaker.error(e);
        }
    }

    public Process getProcess() throws IOException {
        return launcher.Console()
                ? Runtime.getRuntime().exec(new String[]{"cmd.exe", "/C", "start", StartBat.getCanonicalPath()})
                : new ProcessBuilder(StartBat.getCanonicalPath()).directory(launcher.getVersionPath()).start();

    }

    @SneakyThrows(IOException.class)
    public void writeStartScript() {
        StringBuilder Script = new StringBuilder().append(new GameJvmCommand(launcher).getCommand())
                .append(new GameCLICommand(launcher).getCommand());
        logmaker.info(Script);
        FileUtils.writeStringToFile(FileManger.getStarterBat(), Script.toString());
    }
}

File: core/core-impl/src/main/kotlin/org/wdt/wdtc/core/impl/launch/LaunchGame.kt
package org.wdt.wdtc.core.impl.launch

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.launch
import org.wdt.utils.io.isFileNotExists
import org.wdt.utils.io.writeStringToFile
import org.wdt.wdtc.core.impl.download.InstallGameVersion
import org.wdt.wdtc.core.openapi.game.Version
import org.wdt.wdtc.core.openapi.manger.*
import org.wdt.wdtc.core.openapi.utils.*
import java.util.*

class LaunchGame(
	private val version: Version,
	private val coroutineScope: CoroutineScope = ioCoroutineScope
) {
	private val launchTaskProcess: ProcessBuilder
		get() = if (currentSetting.console)
			ProcessBuilder("cmd.exe", "/C", "start", starterBat.canonicalPath).directory(version.versionDirectory)
		else
			ProcessBuilder(starterBat.canonicalPath).directory(version.versionDirectory)
	
	private fun beforLaunchTask() {
		if (!currentSetting.chineseLanguage) return
		version.gameOptionsFile.run {
			if (isFileNotExists()) {
				buildString {
					append("forceUnicodeFont:true").appendLine()
					append("guiScale:2").appendLine()
					append("lang:zh_cn").appendLine()
					append("autoJump:false").appendLine()
				}.also {
					writeStringToFile(it)
				}
			}
		}
	}
	
	// TODO: Optimize lag
	private fun getFullTaskList(
		block: (String) -> Unit = { println(it) }
	): MutableList<TaskManger> = LinkedList<TaskManger>().apply {
		
		add(object : TaskManger("启动前配置", TaskKind.COROUTINES) {
			init {
				coroutinesAction =
					coroutineScope.launch("before launch task".toCoroutineName(), start = CoroutineStart.LAZY) {
						beforLaunchTask()
					}
			}
		})
		
		add(object : TaskManger("补全文件", TaskKind.COROUTINES) {
			init {
				coroutinesAction = coroutineScope.launch("Complete the file".toCoroutineName(), start = CoroutineStart.LAZY) {
					logmaker.info("Start Download")
					InstallGameVersion(version, false).run {
						startInstallGame()
					}
					logmaker.info("Downloaded Finish")
					
				}
			}
		}
		)
		
		add(object : TaskManger("写入启动脚本", TaskKind.COROUTINES) {
			init {
				coroutinesAction =
					coroutineScope.launch("Write start script".toCoroutineName(), start = CoroutineStart.LAZY) {
						logmaker.info("Write Start Script")
						val command = buildString {
							append(GameJvmCommand(version).getCommand())
							append(GameCLICommand(version).getCommand())
						}.also {
							starterBat.writeStringToFile(it)
						}
						logmaker.info(command)
						logmaker.info("Write Start Script Finish")
						
					}
			}
		})
		add(object : TaskManger("运行脚本", TaskKind.COROUTINES) {
			init {
				coroutinesAction = coroutineScope.launch("Run script".toCoroutineName(), start = CoroutineStart.LAZY) {
					logmaker.info("Start Run Start Script,Console:${currentSetting.console}")
					version.run {
						logmaker.info("Launch Version: $versionNumber - $kind")
						logmaker.info(gameConfig.config)
					}
					LaunchProcess(launchTaskProcess, block).run {
						startLaunchGame()
					}
				}
			}
		})
	}
	
	private fun getSimpleTaskList(block: (String) -> Unit): MutableList<TaskManger> {
		return mutableListOf(TaskManger("启动游戏", TaskKind.COROUTINES).apply {
			coroutinesAction = defaultCoroutineScope.launch("Run all game".toCoroutineName(), CoroutineStart.LAZY) {
				getFullTaskList(block).forEach {
					it.coroutinesAction.noNull().run {
						start()
						join()
					}
				}
			}
		})
	}
	
	fun getTaskList(block: (String) -> Unit): MutableList<TaskManger> {
		return if (isLowMode) getSimpleTaskList(block) else getFullTaskList(block)
	}
	
	companion object {
		val Version.launchGameTask
			get() = LaunchGame(this)
	}
}
